### PR TITLE
refactor: self-containing classes

### PR DESCRIPTION
**Pull request by saschanaz**
A trial to implement https://github.com/w3c/webidl2.js/issues/138#issuecomment-455918738.

### CODE REVIEW

**Comment by marcoscaceres** [comment]
Looking good.

**Comment by saschanaz** [comment]
No real value here yet, the commit here is for future self-containing class instances get their own token references that include position information.

TIL new GitHub draft PR so I wanted to try it 😆, IMO good as this will constantly remind me that I have a WIP branch.

**Comment by marcoscaceres** [comment]
How does one do the Draft PR? That’s quite a useful feature.

**Comment by saschanaz** [comment]
It's a fancy new feature: https://github.blog/2019-02-14-introducing-draft-pull-requests/

**Comment by saschanaz** [comment]
IMO mostly done, not feel great about `toJSON()` but it's needed for JSON compatibility.

**Comment by saschanaz** [comment]
Too much changes so I'll open smaller PRs soon.

**Comment by marcoscaceres** [comment]
Appreciate that. Maybe we can do one class at a time? Generally ok if it's around ~100 per pull request.

**Comment by saschanaz** [comment]
@marcoscaceres This still have two pieces of features: name duplication checker and a feature for https://github.com/KhronosGroup/WebGL/issues/2216. Should I split them?

**Comment by marcoscaceres** [comment]
This is manageable now, so no need to split them. I'll try to review tonight or first thing tomorrow.

File: lib/validator.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+    function lastLine(text) {
+      const splitted = text.split("\n");
+      return splitted[splitted.length - 1];
+    }
+  }
+
+  function groupDefinitions(all) {
+    const unique = new Map();
+    const duplications = new Set();
```
Comment (marcoscaceres) [inline lib/validator.js]: Nit: “duplicates”?

Comment (marcoscaceres) [reply lib/validator.js]: Actually, it’s fine to leave it. Would mean changing this in way too many places.

File: lib/validator.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      ...checkDuplicatedNames(defs),
+      ...checkInterfaceMemberDuplication(defs)
+    ];
+  }
+
+  const obj = {
+    validate
+  };
+
+  if (typeof module !== 'undefined' && typeof module.exports !== 'undefined') {
```
Comment (marcoscaceres) [inline lib/validator.js]: I wonder if we should just ship the ES6 Module and not UMD this? I guess we’d still need to bundle it for distribution but might be nice to ship just ES6.

File: lib/validator.js (hunk 1 above)
Comment (saschanaz) [reply lib/validator.js]: That's why IDEs have variable renaming feature 😁

File: lib/validator.js (hunk 2 above)
Comment (saschanaz) [reply lib/validator.js]: I'd like to do that, will open a PR adding webpack.

File: lib/validator.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
+"use strict";
+
+(() => {
+  function error(source, token, current, str) {
+    const consume_position = token.index;
+    const maxTokens = 5;
```
[the remark this answers is not in the document]
Comment (saschanaz) [reply lib/validator.js]: It's rather arbitrary. Hmm.

[the remark this answers is not in the document]
Comment (marcoscaceres) [reply lib/validator.js]: Yeah, that’s what I was wondering. It’s ok for it to be, but just sayin “it works well enough“ is fine.

**Pull request merged**